Mathlib/Data/Nat/Factorial/SuperFactorial.lean

```diff
@@ -78,6 +82,69 @@ theorem det_vandermonde_id_eq_superFactorial (n : ℕ) :
       simp [Fin.prod_univ_eq_prod_range (fun i ↦ (↑i + 1)) (n + 1)]
     · rw [Matrix.det_vandermonde] at hn
       simp [hn]
+open Finset
+
+theorem superFactorial_eq_square_times_factorial (k : ℕ) :
+    ∃ n, IsSquare n ∧ sf (4 * k) = n * (2 * k) ! := by
+  by_cases k = 0
+  · exact ⟨1, ⟨isSquare_one, h ▸ rfl⟩⟩
+  have hk := Nat.sub_add_cancel <| succ_mul_pos 1 <| Nat.pos_of_ne_zero h
+  have h_succ : succ (2 * k - 1) = 2 * k := by linarith
+  use (∏ x in Ico 0 (2 * k - 1), (x + 1) !) * (∏ x in Ico (2 * k) (4 * k), (x + 1) !)
+  have h : sf (4 * k) = (∏ x in Ico 0 (2 * k - 1), (x + 1) !) *
+       (∏ x in Ico (2 * k) (4 * k), (x + 1) !) * (2 * k) ! := by
+    rw [mul_assoc, mul_comm _ (2 * k) !, prod_Ico_factorial_add_one,
+        ← mul_assoc, mul_comm _ (2 * k)!]
+    have := superFactorial_succ (2 * k - 1)
+    simp only [hk] at this
+    norm_num at this
+    rw [← this]
+    rw [← prod_Ico_factorial_add_one, ← prod_Ico_factorial_add_one, h_succ]
+    exact (Finset.prod_Ico_consecutive (fun x => (x + 1) !) (zero_le (2 * k)) (by linarith)).symm
+  constructor
+  · have h' : sf (4 * k) =
+       (2 ^ k * ∏ x in Ico 0 (2 * k), (2 * x + 1)!) ^ 2 * (2 * k)! := by
+      calc
+        sf (4 * k) = ∏ x in Ico 0 (2 * (2 * k)), (x + 1) ! := by
+          rw [←prod_Ico_factorial_add_one, ← mul_assoc]
+        _ = ∏ x in Ico 0 (2 * k), ((2 * x + 1) !) * (((1 + 2 * x) + 1) !) := by
+          rw [mul_comm]
+          have prod_even_oddFin : ∀ (n : ℕ), ∀  (f : ℕ → ℕ),
+              ∏ x : Fin (n * 2), f x =  ∏ x : Fin n, (f (2 * x)) * (f (1 + 2 * x)) := by
+            intros n f
+            rw [(Equiv.prod_comp' finProdFinEquiv (fun i ↦ f ↑(finProdFinEquiv i)) (fun i ↦ f ↑i)
+                    (congrFun rfl)).symm, Fintype.prod_prod_type]
+            simp only [finProdFinEquiv_apply_val]
+            congr
+            ext x
+            rw [Fin.prod_univ_eq_prod_range (fun x_1 => f (↑x_1 + 2 * ↑x)) 2,
+                Finset.prod_range_succ]
+            simp
+          have prod_even_oddIco : ∀ (n : ℕ), ∀ (f : ℕ → ℕ),
+              ∏ x in Ico 0 (n * 2), f x =  ∏ x in Ico 0 n, (f (2 * x)) * (f (1 + 2 * x)) := by
+            intros n f
+            convert prod_even_oddFin n f
+            · rw [Ico_zero_eq_range, ← (Fin.prod_univ_eq_prod_range f (n * 2))]
+            · rw [Ico_zero_eq_range, ← (Fin.prod_univ_eq_prod_range _ n)]
```

Comment:
I'd be happy to outsource these two lemmas, see my question here: 
https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/group.20sum.2Fprod.20in.20pairs.20even.20and.20odd.20summands

Reply:
Why not just prove (not tested):
```lean
@[to_additive]
lemma Finset.prod_even_odd_range {M : Type*} [CommMonoid M] (n : Nat) (f : Nat → M) :
    (∏ x in range n, f (2 * x)) * (∏ x in range n, f (2 * x + 1)) = ∏ x in range (2 * n), f x :=
```
by induction on `n`?

Reply:
That would definitely work, see my very first message in the zulip link above. 

Do you propose to just change it to use induction and not to think about generalizing this any more?

Reply:
I pushed a proof by induction (also, it provides an explicit formula for `n`). If you want to restore helper lemmas from your old proof, then I suggest that you open a new PR with them.
I guess, we need a review from another maintainer now.
maintainer merge